### PR TITLE
New alloc feature

### PR DESCRIPTION
**Pull request by devrandom**
Allows use of `Secp256k1::new` and related API, if an allocator is available

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -135,6 +135,10 @@ pub use secp256k1_sys as ffi;
 #[cfg(any(test, feature = "rand"))] use rand::Rng;
 #[cfg(any(test, feature = "std"))] extern crate core;
 #[cfg(all(test, target_arch = "wasm32"))] extern crate wasm_bindgen_test;
+#[cfg(feature = "alloc")] extern crate alloc;
+
+#[cfg(all(feature = "std", feature = "alloc"))]
+compile_error!("std and alloc are mutually exclusive");
```
Comment (RCasatta) [inline src/lib.rs:141]: I've done the same in another PR https://github.com/rust-bitcoin/bitcoin_hashes/pull/127

however, this causes an issue with how current resolver threat dependencies, taking the union of features if different crates specify different features. New resolver fixes this issue https://blog.rust-lang.org/2021/03/25/Rust-1.51.0.html#cargos-new-feature-resolver but it's much more modern than our MSRV

Comment (devrandom) [reply src/lib.rs:141]: so are you thinking that if they are both set, then we should act as if `std` only is set?

Comment (devrandom) [reply src/lib.rs]: added a fixup for this

**Review by TheBlueMatt — APPROVED** [review]

**Comment by apoelstra** [comment]
needs rebase

**Comment by devrandom** [comment]
rebased

**Review by TheBlueMatt — APPROVED** [review]

**Review by real-or-random — APPROVED** [review]

**Review by elichai — APPROVED** [review]

**Comment by elichai** [comment]
Reminder that when our MSRV hits 1.36 we could simplify and remove a bunch of these cfg conditions and use `alloc` by default
